File: giftbox/sol/sol.cpp
#include <cstdlib>
#include <cstdint>
#include <cstdio>
#include <cmath>
#include <cstring>
#include <string>
#include <iostream>

using namespace std;

double points[500][2];

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.precision(10);

    int n;
    cin >> n;

    for (int i = 0; i < n; ++i) {
        cin >> points[i][0] >> points[i][1];
    }

    double answer = INFINITY;

    // try each rotation angle
    // maximum error = 4 * 2000 * sin(step / 2)
    // allowed error = 0.1
    // step = arcsin(0.1 / 4 / 2000) * 2 ~= 1 / 40000
    for (double a = 0; a < 1.5708; a += 1. / 40000) {
        double ksin = sin(a);
        double kcos = cos(a);
        double xmin = INFINITY;
        double xmax = -INFINITY;
        double ymin = INFINITY;
        double ymax = -INFINITY;

        // the perimeter is decided by the left/right/top/bottom-most points
        for (int i = 0; i < n; ++i) {
            double x = points[i][0] * kcos - points[i][1] * ksin;
            double y = points[i][0] * ksin + points[i][1] * kcos;

            xmin = min(xmin, x);
            xmax = max(xmax, x);
            ymin = min(ymin, y);
            ymax = max(ymax, y);
        }

        // update the answer
        answer = min(answer, 2 * (xmax - xmin + ymax - ymin));
    }

    cout << round(answer) << endl;

    return 0;
}

File: giftbox/sol/sol.py
import math

n = int(input())
points = [
    [
        float(part)
        for part in input().split(' ')
    ]
    for _ in range(n)
]

def test(a):
    ksin = math.sin(a)
    kcos = math.cos(a)
    xmin = math.inf
    xmax = -math.inf
    ymin = math.inf
    ymax = -math.inf

    for point in points:
        x = point[0] * kcos - point[1] * ksin
        y = point[0] * ksin + point[1] * kcos
        xmin = min(xmin, x)
        xmax = max(xmax, x)
        ymin = min(ymin, y)
        ymax = max(ymax, y)

    return 2 * (xmax - xmin + ymax - ymin)

a_best = 0
c_best = math.inf

# Get extrememly close to awnser
for i in range(1000):
    a = math.pi * i / 2000
    c = test(a)

    if c_best > c:
        a_best = a
        c_best = c

# Get even closer to awnser
for i in range(200):
    a = a_best + math.pi * (i - 100) / 200000
    c_best = min(c_best, test(a))

print(round(c_best))
